lib/valueptr.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Cppcheck - A tool for static C/C++ code analysis
+ * Copyright (C) 2007-2019 Cppcheck team.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef GUARD_VALUEPTR_HPP
+#define GUARD_VALUEPTR_HPP
+
+#include "config.h"
+#include <functional>
+#include <memory>
+
+template <class T>
```

Comment:
We need to copy the `ForwardAnalyzer` but since we use inheritance we cant copy the object directly(as this will cause slicing). Instead, we use this class to store the derived copy constructor. This allows us to essentially pass `ForwardAnalyzer` by value(although we are actually passing `ValuePtr<ForwardAnalyzer>`).

Reply:
So `ForwardAnalyzer` could have a virtual method `clone()` that must be implemented in every derived class and that returns a copy.

Reply:
First, there is no way to ensure the `clone` method is implemented for all derived classes. Right now, there is only one class, but I plan to add more classes in the future which may also deepen the inheritance hierarchy(and `clone` methods in the leaf classes could easily be forgotten).

Secondly, the copy for `ForwardAnalyzer` is not called directly. Instead, it copies `ForwardTraversal`, which copies `ForwardAnalyzer`. 

Thirdly, [`copyable`](https://en.cppreference.com/w/cpp/concepts/copyable) is clearly defined by C++ and it does not include calling a `clone` method. So copying using `clone` will be unfamiliar to most C++ programmers(unless they came from java). Furthermore, if we ever want to use these classes with C++ algorithms or containers than using non-standard copying won't work.

`ValuePtr` is a fairly simple and lightweight class that improves correctness, composability, and simplicity.

Reply:
Well I guess it's more productive for us if I choose a battle that matters more. Let's have this your way.